xla/pjrt/gpu/se_gpu_pjrt_client.cc

```diff
@@ -778,13 +779,17 @@ GetStreamExecutorGpuDeviceAllocator(
 
   // Add any additional allocators for alternate memory spaces.
   if (allocator_config.collective_memory_size != 0) {
+    bool preallocate;
+    TF_RETURN_IF_ERROR(tsl::ReadBoolFromEnvVar(
+        "TF_COLLECTIVE_MEMORY_PREALLOCATE", false, &preallocate));
+
     for (const auto& ordinal_and_device : addressable_devices) {
       TF_ASSIGN_OR_RETURN(
           auto collective_bfc_allocator,
           CreateCollectiveBFCAllocator(
               ordinal_and_device.second->executor(),
               /*allocator_memory=*/allocator_config.collective_memory_size,
```

Comment:
Shouldn't we also subtract collective_memory_size from device available memory above at the BFC allocator in case of memory preallocation?
Currently we are relying on a fact that available device memory * allocator_config.memory_fraction > collective memory, which is a bit fragile.

Reply:
I think this makes sense, because the buffers which are now going to the collective BFC pool previously were part of the standard BFC pool. This means we are not really reducing the memory available to the users model if the `XLA_PYTHON_CLIENT_COLLECTIVE_MEM_SIZE_MB` is tuned properly. However, if that value is set way higher than the user needs, they will end up starving the standard allocator which might cause more OOMs.